### PR TITLE
feature_importance_visualization for multiclass problems

### PR DESCRIPTION
**Pull request by chidauri**
Regarding #162

### CODE REVIEW

**Comment by marco-c** [comment]
Can you show an example output?

**Comment by chidauri** [comment]
> Can you show an example output?

Selection_002

File: bugbug/model.py [hunk 1]
```diff
@@ -193,18 +193,40 @@ def train(self, importance_cutoff=0.15):
             shap_values = explainer.shap_values(X_train)
 
             # TODO: Actually implement feature importance visualization for multiclass problems.
```
Comment (marco-c) [inline bugbug/model.py]: You can now remove this TODO, since it is what you are doing now

File: bugbug/model.py [hunk 2]
```diff
@@ -193,18 +193,40 @@ def train(self, importance_cutoff=0.15):
...
+            new_shap_values = 0
```
Comment (marco-c) [inline bugbug/model.py]: I wouldn't call them `new`, `average` would be better

File: bugbug/model.py [hunk 3]
```diff
@@ -193,18 +193,40 @@ def train(self, importance_cutoff=0.15):
...
+                    print(
+                        f'{i + 1}. \'{feature_names[int(index)]}\' ({"+" if (is_positive) else "-"}{importance})'
+                    )
+            else:
+                print("{:<}".format("classes"), end="   ")
+                for importance, index, is_pos in important_features:
+                    print("{:>}".format(feature_names[int(index)]), end="  ")
+                print("\n")
+                for class_num, item in enumerate(shap_values):
+                    print("class {}".format(class_num), end="   ")
```
Comment (marco-c) [inline bugbug/model.py]: We should print the class name rather than the class number

Comment (chidauri) [reply bugbug/model.py]: I've assumed the classes be in ascending order of their names, i don't know if it is correct :p

**Comment by marco-c** [comment]
You'll need to rebase on top of master, as there have been some changes around the same lines.
Can you post another example output with the recent changes?

**Comment by marco-c** [comment]
Oh, and you'll have to implement the same in the `classify` function.

**Comment by chidauri** [comment]
> Oh, and you'll have to implement the same in the `classify` function.

The print functionality is in run.py, so for printing multi-class examples we'll have to return class-names as well as shap_values, so can we move the printing in model.py in the classify function itself?

**Comment by marco-c** [comment]
In the classify case, we are only interested in the most important features for the predicted class (the user wants to know "why did we select this class?")

File: bugbug/model.py [hunk 4]
```diff
@@ -192,19 +192,45 @@ def train(self, importance_cutoff=0.15):
...
-                importance_cutoff, shap_values
+                importance_cutoff, avg_shap_values
             )
 
-            print(f"\nTop {len(important_features)} Features:")
-            for i, [importance, index, is_positive] in enumerate(important_features):
-                print(
-                    f'{i + 1}. \'{feature_names[int(index)]}\' ({"+" if (is_positive) else "-"}{importance})'
-                )
+            if not isinstance(shap_values, list):
```
Comment (marco-c) [inline bugbug/model.py]: I'd suggest refactoring this code to use tabulate, this would make it much cleaner.

Comment (chidauri) [reply bugbug/model.py]: you mean, same as multi class make it  horizontal instead of vertical ?

Comment (marco-c) [reply bugbug/model.py]: I meant using `tabulate`, which is a Python package to pretty-print tabular data such as this (both in the binary and multi-class cases).

**Comment by chidauri** [comment]
@marco-c 
Selection_003
after using tabulate(in the default 'simple' table format)

File: bugbug/model.py (hunk 3 above)
Comment (chidauri) [reply bugbug/model.py]: @marco-c this should be ascending or descending? cause you have used descending order in summary plot

**Comment by chidauri** [comment]
@marco-c  After recent changes.

Selection_005
This is how binary classification looks.
Selection_006
Multi class classification.


PS. We don't need "is_pos" parameter from get_important_features method anymore.

File: bugbug/model.py [hunk 5]
```diff
@@ -203,46 +204,40 @@ def train(self, importance_cutoff=0.15):
             )
 
             matplotlib.pyplot.savefig("feature_importance.png", bbox_inches="tight")
-            
+
```
Comment (chidauri) [inline bugbug/model.py]: i will fix this :p

**Comment by marco-c** [comment]
The binary classification table doesn't look great, maybe there is a way for tabulate to handle better the large number of columns?

File: bugbug/model.py (hunk 3 above)
Comment (marco-c) [reply bugbug/model.py]: Not sure, we need to confirm with a real case and see if the results make sense in ascending or in descending order.

File: bugbug/model.py [hunk 6]
```diff
@@ -204,20 +205,40 @@ def train(self, importance_cutoff=0.15):
...
             print(f"\nTop {len(important_features)} Features:")
-            for i, [importance, index, is_positive] in enumerate(important_features):
-                print(
-                    f'{i + 1}. \'{feature_names[int(index)]}\' ({"+" if (is_positive) else "-"}{importance})'
+            top_indexes = list(
+                int(index) for importance, index, is_pos in important_features
+            )
+
+            table = []
+            for num, item in enumerate(shap_values):
```
Comment (marco-c) [inline bugbug/model.py]: Part of this should be moved to get_important_features.
get_important_features could return an object like:
```
{
    "average": most important features in average shap values,
    "classes": {
        "class1": (most important features in class1 shap values, values in class1 of the most important features in average shap values),
        "class2": (most important features in class1 shap values, values in class1 of the most important features in average shap values),
    }
}
```

This way here you can show the most important feaures in average shap values and their values for each class, and in classify you can show the most important features for the predicted class.

Comment (chidauri) [reply bugbug/model.py]: so we want the list of most important features for each class, distinctly, to use in classify but we do not need that here, right?

Comment (marco-c) [reply bugbug/model.py]: Exactly.
At training time, we will show the most important average features and their values for all classes.
At classify time, we will show the most important features for the predicted class.

get_important_features will return everything though, so we can easily do more complex stuff in the future if we want to.

**Comment by chidauri** [comment]
> The binary classification table doesn't look great, maybe there is a way for tabulate to handle better the large number of columns?

we can use transpose(features  as rows and classes as columns) for both cases as features are always going to be greater than number of classes so we won't have to worry if features increase in multi class case too.

**Comment by marco-c** [comment]
> we can use transpose(features as rows and classes as columns) for both cases as features are always going to be greater than number of classes so we won't have to worry if features increase in multi class case too.

Unfortunately in the "component" model we have a lot of classes :(

**Comment by chidauri** [comment]
@marco-c
Binary classification. 
Selection_007
multi-class classify function:
Selection_008
mutli-class classification:
Selection_006

File: bugbug/model.py [hunk 7]
```diff
@@ -113,11 +114,14 @@ def le(self):
     def get_feature_names(self):
         return []
 
-    def get_important_features(self, cutoff, shap_values):
+    def get_important_features(
+        self, cutoff, avg_shap_values, shap_values, class_names=False
```
Comment (marco-c) [inline bugbug/model.py]: Move the avg calculation in this function, so you can only pass `shap_values` to it

File: bugbug/model.py [hunk 8]
```diff
@@ -133,7 +137,40 @@ def get_important_features(self, cutoff, shap_values):
...
+                rel_sums = abs_sums / abs_sums.sum()
+                is_pos = ["+" if shap_sum >= 0 else "-" for shap_sum in item.sum(0)]
+                for index in important_features["average"]:
+                    top_avg.append(is_pos[index] + str(rel_sums[index]))
+
+                if class_names:
+                    important_features["classes"][
+                        class_names[len(class_names) - num - 1]
+                    ] = (top_item_features, top_avg)
+                else:
```
Comment (marco-c) [inline bugbug/model.py]: In which case do you not have class names?

File: bugbug/model.py [hunk 9]
```diff
@@ -133,7 +137,40 @@ def get_important_features(self, cutoff, shap_values):
...
+                top_avg = []
+                abs_sums = np.abs(item).sum(0)
+                rel_sums = abs_sums / abs_sums.sum()
+                is_pos = ["+" if shap_sum >= 0 else "-" for shap_sum in item.sum(0)]
+                for index in important_features["average"]:
+                    top_avg.append(is_pos[index] + str(rel_sums[index]))
+
+                if class_names:
+                    important_features["classes"][
+                        class_names[len(class_names) - num - 1]
```
Comment (marco-c) [inline bugbug/model.py]: Are you sure they are in reverse order?

File: bugbug/model.py [hunk 10]
```diff
@@ -133,7 +137,40 @@ def get_important_features(self, cutoff, shap_values):
...
+                    cutoff, item, item, class_names
+                )
+
+                # top features in average shap values for that class
+                top_avg = []
+                abs_sums = np.abs(item).sum(0)
+                rel_sums = abs_sums / abs_sums.sum()
+                is_pos = ["+" if shap_sum >= 0 else "-" for shap_sum in item.sum(0)]
+                for index in important_features["average"]:
+                    top_avg.append(is_pos[index] + str(rel_sums[index]))
```
Comment (marco-c) [inline bugbug/model.py]: You can use a list comprehension instead of iterating and appending

File: bugbug/model.py [hunk 11]
```diff
@@ -133,7 +137,40 @@ def get_important_features(self, cutoff, shap_values):
         # Sort the array (in decreasing order of importance values)
         top_features = top_features[top_features[:, 0].argsort()][::-1]
 
-        return top_features
+        if isinstance(shap_values, list):
+            important_features = {}
+            important_features["classes"] = {}
+            important_features["average"] = list(
```
Comment (marco-c) [inline bugbug/model.py]: We should store `top_features` in here

File: bugbug/model.py [hunk 12]
```diff
@@ -204,18 +241,34 @@ def train(self, importance_cutoff=0.15):
...
             )
 
-            print(f"\nTop {len(important_features)} Features:")
-            for i, [importance, index, is_positive] in enumerate(important_features):
+            print("Top {} features:".format(len(important_features["average"])))
+            top_feature_names = [
+                feature_names[index] for index in important_features["average"]
+            ]
+
+            # allow maximum of 8 columns in a row to fit the page better
```
Comment (marco-c) [inline bugbug/model.py]: We should share as much as possible this code between model.py and run.py. You can add a new function "print_feature_importances".

File: bugbug/model.py (hunk 9 above)
Comment (chidauri) [reply bugbug/model.py]: Actually not in reverse order but currently they are already reversed in model.py so i reversed them again.

File: bugbug/model.py (hunk 8 above)
Comment (chidauri) [reply bugbug/model.py]: in classify case ??

Comment (marco-c) [reply bugbug/model.py]: We should define class names as part of the models, so you can use it here

**Comment by chidauri** [comment]
@marco-c  after the recent commit , printing is working fine for training , but will have to fix it for classify.
These are the results for train:
defect train:
defect_train
defect_enhancement_task train:
defectenhancetask_train

File: bugbug/model.py [hunk 13]
```diff
@@ -204,19 +288,14 @@ def train(self, importance_cutoff=0.15):
 
             matplotlib.pyplot.savefig("feature_importance.png", bbox_inches="tight")
 
-            # TODO: Actually implement feature importance visualization for multiclass problems.
-            if isinstance(shap_values, list):
-                shap_values = np.sum(np.abs(shap_values), axis=0)
+            if not isinstance(shap_values, list):
```
Comment (marco-c) [inline bugbug/model.py]: Move this to get_important_features

File: bugbug/model.py [hunk 14]
```diff
@@ -300,16 +379,19 @@ def classify(
             explainer = shap.TreeExplainer(self.clf)
             shap_values = explainer.shap_values(X)
 
-            # TODO: Actually implement feature importance visualization for multiclass problems.
-            if isinstance(shap_values, list):
-                shap_values = np.sum(np.abs(shap_values), axis=0)
+            if not isinstance(shap_values, list):
```
Comment (marco-c) [inline bugbug/model.py]: Here too

File: run.py [hunk 15]
```diff
@@ -103,12 +103,14 @@ def main(args):
...
                 feature_names = model.get_feature_names()
-                for i, (importance, index, is_positive) in enumerate(
-                    importance["importances"]
-                ):
-                    print(
-                        f'{i + 1}. \'{feature_names[int(index)]}\' ({"+" if (is_positive) else "-"}{importance})'
-                    )
+
+                # shap values of top features for the predicted class
+                pred_class_index = int(np.where(probas == np.amax(probas))[1])
```
Comment (marco-c) [inline run.py]: The index should just be probas.argmax(axis=-1)[0]

File: bugbug/model.py (hunk 13 above)
Comment (chidauri) [reply bugbug/model.py]: if we move this to the function itself , we will have to write separate code for single class and multiclass, i mean how will the recursion work?

File: run.py [hunk 16]
```diff
@@ -103,12 +103,14 @@ def main(args):
...
+
+                model.print_feature_importances(
+                    importance["importances"], feature_names, pred_class_index
```
Comment (chidauri) [inline run.py]: @marco-c i need help here , when i call this (even with pred_class = pred_class_index), pred_class is being treated as False(the default values) :(

Comment (chidauri) [reply run.py]: i got it.

File: bugbug/model.py [hunk 17]
```diff
@@ -386,11 +386,17 @@ def classify(
                 importance_cutoff, shap_values
             )
 
-            # TODO: pred class index
+            # Workaround: handle multi class case for force_plot to work correctly
+            if not len(classes[0]) == 2:
+                pred_class_index = classes.argmax(axis=-1)[0]
+                explainer.expected_value = explainer.expected_value[0]
+            else:
+                pred_class_index = 0
```
Comment (chidauri) [inline bugbug/model.py]: @marco-c for binary case , we have values stored in  a single class i.e., ["class 0"], so we have to make the index 0, and not the one with maximum probas.

File: bugbug/model.py [hunk 18]
```diff
@@ -300,24 +379,33 @@ def classify(
...
+                ][0]
             ]
 
             feature_names = self.get_feature_names()
 
             with io.StringIO() as out:
                 p = shap.force_plot(
                     explainer.expected_value,
-                    shap_values[:, top_indexes],
+                    shap_values[0][:, top_indexes],
```
Comment (marco-c) [inline bugbug/model.py]: Shouldn't this be shap_values[pred_class_index]?

File: bugbug/model.py [hunk 19]
```diff
@@ -329,7 +417,14 @@ def classify(
 
                 html = out.getvalue()
 
-            return classes, {"importances": top_importances, "html": html}
+            return (
+                classes,
+                {
+                    "importances": important_features,
+                    "html": html,
+                    "predicted_class": pred_class_index,
```
Comment (marco-c) [inline bugbug/model.py]: This doesn't seem useful, the caller should be able to get the pred_class_index from `classes`

File: bugbug/model.py [hunk 20]
```diff
@@ -300,24 +379,33 @@ def classify(
...
+            if not isinstance(shap_values, list):
+                shap_values = [shap_values]
 
-            top_importances = self.get_important_features(
+            important_features = self.get_important_features(
                 importance_cutoff, shap_values
             )
 
+            # Workaround: handle multi class case for force_plot to work correctly
+            if not len(classes[0]) == 2:
```
Comment (marco-c) [inline bugbug/model.py]: if len(classes[0]) != 2

File: bugbug/model.py [hunk 21]
```diff
@@ -21,6 +21,7 @@
 from sklearn.feature_extraction.text import TfidfVectorizer
 from sklearn.metrics.classification import precision_recall_fscore_support
 from sklearn.model_selection import cross_validate, train_test_split
+from tabulate import tabulate
```
Comment (marco-c) [inline bugbug/model.py]: You'll need to add this to the requirements.txt file (see the failing tests)

File: bugbug/model.py [hunk 22]
```diff
@@ -133,7 +140,84 @@ def get_important_features(self, cutoff, shap_values):
...
+                        top_avg,
+                    )
+
+            return important_features
+
+        else:
+            return top_features
+
+    def print_feature_importances(
+        self, important_features, feature_names, pred_class=False
```
Comment (marco-c) [inline bugbug/model.py]: Nit: rename `pred_class` to `predicted_class`

File: bugbug/model.py (hunk 13 above)
Comment (marco-c) [reply bugbug/model.py]: You can split the function in two, one which calculates and returns top_features for a shap_values matrix, and one which is called with a shap_values list and calls the first one.

File: bugbug/model.py (hunk 18 above)
Comment (chidauri) [reply bugbug/model.py]: Traceback (most recent call last):
  File "run.py", line 165, in <module>
    main(parse_args(sys.argv[1:]))
  File "run.py", line 102, in main
    bug, probabilities=True, importances=True
  File "/home/harshit/Documents/bugbug/bugbug/model.py", line 398, in classify
    shap_values[pred_class_index][:, top_indexes],
IndexError: too many indices for array

File: bugbug/model.py (hunk 21 above)
Comment (chidauri) [reply bugbug/model.py]: added this, but still failing tests :(

File: run.py [hunk 23]
```diff
@@ -103,12 +103,19 @@ def main(args):
...
 
                 feature_names = model.get_human_readable_feature_names()
-                for i, (importance, index, is_positive) in enumerate(
-                    importance["importances"]
-                ):
-                    print(
-                        f'{i + 1}. \'{feature_names[int(index)]}\' ({"+" if (is_positive) else "-"}{importance})'
-                    )
+
+                # shap_values are stored in class 0 for binary classification
```
Comment (marco-c) [inline run.py]: I don't like this part, users of the `print_feature_importance` function should not have to know about this. We should move it to print_feature_importances (making `probas` the parameter of print_feature_importances instead of `predicted_class`)

File: bugbug/model.py [hunk 24]
```diff
@@ -140,27 +141,110 @@ def get_human_readable_feature_names(self):
...
+            abs_sums = np.abs(item).sum(0)
+            rel_sums = abs_sums / abs_sums.sum()
+            is_pos = ["+" if shap_sum >= 0 else "-" for shap_sum in item.sum(0)]
+            top_avg = [
+                is_pos[int(index)] + str(rel_sums[int(index)])
+                for importance, index, is_positive in important_features["average"]
+            ]
 
-        return top_features
+            if class_names:
```
Comment (marco-c) [inline bugbug/model.py]: We should always use class_names

File: run.py [hunk 25]
```diff
@@ -103,12 +103,19 @@ def main(args):
...
+                    pred_class_index = probas.argmax(axis=-1)[0]
+                else:
+                    pred_class_index = 0
+
+                model.print_feature_importances(
+                    importance["importances"],
+                    feature_names,
+                    predicted_class=pred_class_index,
+                )
+
```
Comment (marco-c) [inline run.py]: Nit: don't add newline here

File: run.py [hunk 26]
```diff
@@ -112,12 +112,10 @@ def main(args):
...
                 feature_names = model.get_human_readable_feature_names()
-                for i, (importance, index, is_positive) in enumerate(
-                    importance["importances"]
-                ):
-                    print(
-                        f'{i + 1}. \'{feature_names[int(index)]}\' ({"+" if (is_positive) else "-"}{importance})'
-                    )
+
+                model.print_feature_importances(
+                    importance["importances"], feature_names, predicted_class=probas
```
Comment (marco-c) [inline run.py]: The name of the parameter should change too, since it's no longer `predicted_class`

File: bugbug/model.py [hunk 27]
```diff
@@ -156,27 +156,113 @@ def get_human_readable_feature_names(self):
...
+            # shap values of top average features for that class
+            abs_sums = np.abs(item).sum(0)
+            rel_sums = abs_sums / abs_sums.sum()
+            is_pos = ["+" if shap_sum >= 0 else "-" for shap_sum in item.sum(0)]
+            top_avg = [
+                is_pos[int(index)] + str(rel_sums[int(index)])
+                for importance, index, is_positive in important_features["average"]
+            ]
+
+            important_features["classes"][f"class {class_names[num]}"] = (
```
Comment (marco-c) [inline bugbug/model.py]: Let's directly use the class_name as a key here, instead of "class CLASS_NAME"

File: bugbug/model.py [hunk 28]
```diff
@@ -349,16 +427,23 @@ def classify(
...
             shap_values = explainer.shap_values(X)
 
-            # TODO: Actually implement feature importance visualization for multiclass problems.
-            if isinstance(shap_values, list):
-                shap_values = np.sum(np.abs(shap_values), axis=0)
-
-            top_importances = self.get_important_features(
-                importance_cutoff, shap_values
+            important_features = self.get_important_features(
+                importance_cutoff, shap_values, class_names=range(len(classes[0]))
```
Comment (marco-c) [inline bugbug/model.py]: I've filed #732 to use the real class names here.

File: bugbug/model.py [hunk 29]
```diff
@@ -156,31 +165,116 @@ def get_human_readable_feature_names(self):
...
-        # Get the importance values of the top features from their indices
-        top_features = np.take(rel_shap_sums, top_feature_indices)
-        # Gets the sign of the importance from shap_sums as boolean
-        is_positive = (np.take(shap_sums, top_feature_indices)) >= 0
-        # Stack the importance, indices and shap_sums in a 2D array
-        top_features = np.column_stack((top_features, top_feature_indices, is_positive))
-        # Sort the array (in decreasing order of importance values)
-        top_features = top_features[top_features[:, 0].argsort()][::-1]
+    def print_feature_importances(
+        self, important_features, feature_names, classify=False
```
Comment (marco-c) [inline bugbug/model.py]: `classify` is not a great name, and it's not great that it can be either a boolean or a list, it should be consistent.
You could call it `class_probabilities` and make it None in the non-classify case and a list in the classify case.

File: .isort.cfg [hunk 30]
```diff
@@ -5,3 +5,4 @@ force_grid_wrap=0
 use_parentheses=True
 line_length=88
 known_third_party = apispec,apispec_webframeworks,dateutil,flask,flask_cors,hglib,imblearn,jsone,jsonschema,libmozdata,marshmallow,matplotlib,microannotate,models,numpy,pandas,pkg_resources,pyemd,pytest,redis,requests,responses,rq,setuptools,shap,sklearn,tabulate,taskcluster,tqdm,xgboost,yaml,zstandard
+
```
Comment (marco-c) [inline .isort.cfg]: This change is unrelated to the PR, can you revert it?

File: bugbug/model.py [hunk 31]
```diff
@@ -156,31 +165,116 @@ def get_human_readable_feature_names(self):
 
         return cleaned_feature_names
 
-    def get_important_features(self, cutoff, shap_values):
-        # Calculate the values that represent the fraction of the model output variability attributable
-        # to each feature across the whole dataset.
-        shap_sums = shap_values.sum(0)
-        abs_shap_sums = np.abs(shap_values).sum(0)
-        rel_shap_sums = abs_shap_sums / abs_shap_sums.sum()
+    def get_important_features(self, cutoff, shap_values, class_names):
```
Comment (marco-c) [inline bugbug/model.py]: Since class_names are stored in self, class_names doesn't need to be a parameter of this function. You can just use self.class_names instead of class_names when needed

File: scripts/commit_classifier.py [hunk 32]
```diff
@@ -109,9 +109,15 @@ def classify(self, diff_id):
 
         feature_names = self.model.get_human_readable_feature_names()
 
+        if len(probs[0]) > 2:
```
Comment (marco-c) [inline scripts/commit_classifier.py]: We know that this is a binary case, so we don't need to check

File: scripts/commit_classifier.py [hunk 33]
```diff
@@ -109,9 +109,15 @@ def classify(self, diff_id):
...
+            pred_class_index = probs.argmax(axis=-1)[0]
+        else:
+            pred_class_index = 0
+
+        pred_class = self.model.class_names[pred_class_index]
         features = []
         for i, (val, feature_index, is_positive) in enumerate(
-            importance["importances"]
+            importance["importances"]["classes"][pred_class][0]
```
Comment (marco-c) [inline scripts/commit_classifier.py]: Here too, we know the class name as it's a binary classifier, I think you can just hardcode "1" instead of `pred_class`

File: scripts/commit_classifier.py [hunk 34]
```diff
@@ -121,6 +127,10 @@ def classify(self, diff_id):
                 ]
             )
 
+        self.model.print_feature_importances(
```
Comment (marco-c) [inline scripts/commit_classifier.py]: We don't need to print the features here
